Fix for #7765

stmeyer opened the pull request:
This is a fix for https://github.com/cakephp/cakephp/issues/7765.

Hope I'm doing this well, its my first pull request

jeffblack360 committed:
Add method to return array of listeners matching a pattern

Fix ticket 4192

jeffblack360 committed:
Correct indentation per Travis for multi-line function call

jeffblack360 committed:
Run match pattern through preq_quote()

jeffblack360 committed:
Remove double space to satisfy PHPCS

skie committed:
Added identiferEquals method for expressions to use fields with quoting in join conditions. Added identifier parameter for expressions.

skie committed:
Allow to use the connection name with the option "connectionName" in TableRegistry::get and TableRegistry::config.

skie committed:
Added check for input parameters types in QueryExpression::equalFields.

skie committed:
Coding standart fix.

skie committed:
Remove unneeded elseif

skie committed:
Merge remote-tracking branch 'origin/master' into 3.2-indentifer-equals

skie committed:
Fixed undefined dataset ordering in tests.

markstory committed:
Allow camelBacked virtualProperties to be accessed.

This fixes a regression introduced in while optimizing entity accessors.

Refs #8246

skie committed:
Added link to config documentation for TableRegistry::get

Commit:
fix for https://github.com/cakephp/cakephp/issues/7765

Commit:
fix comment

Commit:
fix phpcs

jadb commented on src/Database/Dialect/SqlserverDialectTrait.php:
```diff
@@ -325,6 +325,10 @@ public function rollbackSavePointSQL($name)
     public function newCompiler()
     {
+        if (isset($this->_config['compiler'])) {
+            $class = $this->_config['compiler'] . 'Compiler';
```
Maybe throw a detailed exception if the compiler is missing (i.e. missing controllers, etc.)? 

Not sure what others think.

jadb commented:
:clap: nice job!

Maybe squash into one commit?

markstory commented on src/Database/SqlserverWithTriggerCompiler.php:
```diff
@@ -0,0 +1,91 @@
+    /**
+     * Generates the INSERT part of a SQL query
+     *
+     * SQL Server with enalbed triggers does not allow the OUTPUT clause
```
Typo in 'enabled'

markstory commented on src/Database/SqlserverWithTriggerCompiler.php:
```diff
@@ -0,0 +1,91 @@
+    {
+        $table = $parts[0];
+        $talbeParts = explode('.', $table);
+        if (count($talbeParts) > 1) {
```
Typo in `tableParts`.

markstory commented on src/Database/SqlserverWithTriggerCompiler.php:
```diff
@@ -0,0 +1,91 @@
+            }
+        }
+        $columns = $this->_stringifyExpressions($parts[1], $generator);
+        return sprintf('DECLARE @var TABLE (%s);INSERT INTO %s (%s) OUTPUT INSERTED.%s INTO @var', implode(', ', $sqlCreateColumns), $table, implode(', ', $columns), implode(' , INSERTED.', $primaryKeyColumnsForSql));
```
Could this really long line be wrapped?

markstory commented on src/Database/SqlserverWithTriggerCompiler.php:
```diff
@@ -0,0 +1,91 @@
+        foreach ($constraints as $constraintName) {
+            $constraint = $description->constraint($constraintName);
+            if ($constraint['type'] === \Cake\Database\Schema\Table::CONSTRAINT_PRIMARY) {
+                $primaryKeyColumns = $constraint['columns'];
```
If you want the primary key columns, the `primaryKey()` method on `Schema\Table` will do just that.

markstory commented:
Having some tests so we can ensure this works now and in the future would be great. Some tests I would like to see are:

1. Test that inserts and updates work.
2. Interact with schema that has autoincrement primary keys
3. Interact with UUID primary keys
4. Interact with composite primary keys. This might overlap with 2.

markstory committed:
Merge pull request #8241 from CakeDC/3.2-table-registry-connection-name-7644

Allow to use the connection name with the option "connectionName" in TableLocator::get()

markstory committed:
Update docs for TableLocator::get().

Make the options documentation more consistent with other methods.

lorenzo committed:
Merge pull request #8242 from CakeDC/3.2-indentifer-equals

Added identiferEquals method for expressions to use fields with quoti…

lorenzo committed:
Fixed random failing tests in travis with php 5.6 and mysql

lorenzo committed:
Merge pull request #8247 from cakephp/issue-8246

Allow camelBacked virtualProperties to be accessed.

stmeyer commented:
@markstory Where can I find comparable tests?

lorenzo commented:
@stmeyer what I would do is to create another testsuite using phpunit and run the core database + orm tests after altering the configuration for sqlserver to use this new compiler.

stmeyer commented:
@lorenzo I never worked with testsuits. I could copy the database testsuit:
```xml
<testsuite name="Database Test Suite">
    <file>./tests/TestCase/DatabaseSuite.php</file>
</testsuite>
```
how to alter the configuration of sqlserver?

ADmad committed:
Account for repeating namespace and nested app namespace.

lorenzo committed:
Fixing bug in the Date::parseDate() method

When using a different timezone it would return different dates
because of the timezone changes.

markstory committed:
Merge pull request #8273 from cakephp/app-shortname

Account for repeating namespace and nested app namespace.

markstory committed:
Honour templatePath set in Cell actions.

When cells set their own templatePath we shouldn't muck about with that.

Refs #8271

lorenzo committed:
Fixing yet another random fail in travis

lorenzo committed:
Fixed mistake from previous commit

lorenzo committed:
Attempting another fix at random fails in travis

ypnos-web committed:
Consistently use formatTemplate()

`formatTemplate()` is a method that can be overwritten by plugins. It is also a method that can be called by the user to manually build customized form elements (e.g. adding extra elements to a `submitContainer`).

Right now FormHelper inconsistently uses `formatTemplate()` or `$this->templater()->format()`. It should in all cases use `formatTemplate()` however, as to provide the same experience to the user in both cases (internal or external use).
Furthermore, overwriting `formatTemplate()` is a powerful tool for plugins, only if it is also consistently called internally.

lorenzo committed:
Un-skipping some tests in order to figure out where the problem comes from

lorenzo committed:
Merge pull request #8275 from cakephp/fix-date-marshal-timezone

Fixing bug in the Date::parseDate() method

lorenzo committed:
Merge pull request #8280 from ypnos-web/patch-3

Consistently use formatTemplate()

lorenzo committed:
Merge pull request #8278 from cakephp/issue-8271

Honour templatePath set in Cell actions.

jeffblack360 committed:
Work on unit-tests for matchingListeners

Fix ticket 4192

Commit:
Cleaned a few more doc blocks

Commit:
Fix some more doc block issues.

markstory committed:
Merge pull request #8283 from cakephp/master-doc-block-fixes

Add fully qualified namespaces to API doc links

Commit:
Adding a line to the docs of Hash and a test case.

lorenzo committed:
Merge pull request #8288 from burzum/hash-docs

Adding a line to the docs of Hash and a test case.

markstory committed:
Merge pull request #8104 from jeffblack360/ticket-4192

Add method to return array of listeners matching a pattern.

Fixes #4192

lorenzo commented:
@stmeyer I think I can help you if you give me commit access to your fork, so that I can add commits on top of yours

stmeyer commented:
@lorenzo I have added you as a collaborator. Thanks for help!

Commit:
Fix sorting of `DateTimeImmutable` instances.

markstory committed:
Merge pull request #8293 from ndm2/fix-immutable-datetime-sort

Fix sorting of `DateTimeImmutable` instances.

markstory committed:
Add Security::random()

This method gives a safer alternative to Text::uuid() when making CSRF
tokens. This method does fallback to an insecure source of data when we
can't find a suitable one. This is only present to preserve backwards
compatbility for folks who can upgrade their servers or take on the
recommended polyfill library.

Refs #8282

ADmad committed:
Allow passing custom finder options to AuthComponent config.

Closes #8295

ADmad committed:
Update docblocks.

ADmad committed:
Move CONTRIBUTIONG.md under .github folder.

ADmad committed:
Add draft templates for github issues and pull requests.

phillaf committed:
Looping through themePaths by counting its own elements

markstory committed:
Correct length issues on fallback code.

The insecure codepath should return the correct length as well.

ADmad committed:
Merge pull request #8299 from cakephp/auth-finder-options

Allow passing custom finder options to AuthComponent config.

ADmad committed:
Update github issues template.

phillaf committed:
Test path generation with multiple paths in App.paths.templates

ADmad committed:
Merge pull request #8300 from cakephp/github-templates

GitHub templates

ADmad committed:
Try to avoid issues being used to ask questions.

ionas committed:
Enable multi lines

ADmad committed:
Merge pull request #8304 from ionas/patch-1

Enable multi lines

antograssiot committed:
remove all reference to DS contant

lorenzo committed:
Merge pull request #8301 from Phillaf/bugfix-count

Looping through themePaths by counting its own elements

markstory committed:
Merge pull request #8305 from cakephp/cleanup-DS-usage

remove all reference to DS contant

lorenzo committed:
Avoiding possible key collisions in the Paginator helper

It was possiblle before to pass keys in the query string that collided
with parts of the url params in the router

lorenzo committed:
Added another test to the router with query string

lorenzo committed:
Merge pull request #8297 from cakephp/security-random

Add Security::randomBytes()

lorenzo committed:
Fixing some issues from previous commits

lorenzo committed:
Update version number to 3.2.3

lorenzo committed:
Changing the publish release step order.

Doing this step lasts helps packagist notice the change and
prevents us from publishing the previous branch

markstory committed:
Merge pull request #8306 from cakephp/release-change

Changing the publish release step order.

antograssiot committed:
remove usage of deprecated ``type()`` method

antograssiot committed:
parameters number mismatch in function call

antograssiot committed:
this tests is not checking what it should

lorenzo committed:
Merge pull request #8312 from antograssiot/master

3.2 cleanup

lorenzo committed:
Fixed CS issue

antograssiot committed:
cleanup imports

ADmad committed:
Remove unneeded headings.

mortonfox committed:
Update link to CONTRIBUTING.md

markstory committed:
Merge pull request #8315 from mortonfox/patch-1

Update link to CONTRIBUTING.md

markstory committed:
Merge pull request #8303 from cakephp/ADmad-patch-1

Try to avoid issues being used to ask questions.

markstory committed:
Add Response::getFile()

Right now if you need to get the file from inside the response, you have
to use reflection tricks. Which makes me sad as I try to build out the
PSR7 bridge for CakePHP. Having a method will be much nicer.

antograssiot committed:
fix the parameter namespace

ADmad committed:
Merge pull request #8320 from cakephp/cleanup

3.2 remove extra use statetements

Commit:
fix typos, and wrapp long line

markstory committed:
Revert "Make find(list) not auto-null on a behavior of 2.x"

lorenzo committed:
Merge pull request #8324 from cakephp/revert-8098-master-find-list

Revert "Make find(list) not auto-null on a behavior of 2.x"

markstory committed:
Make tests pass on windows.

markstory committed:
Quote table name in SQLServer schema.

This allows table names with schema to work correctly.

Refs #8313

adamroyle committed:
Add getter for protected $_className value for Association.

adamroyle committed:
Add tests for new className() method on Association.

markstory committed:
Make JsonConfig pretty print config files.

This makes JSON config files slightly less horrible to work with as you
can read them now.

Refs #8338

markstory committed:
Make route patterns allow multibyte sub-patterns.

This allows route parameter matching patterns to contain multibyte
expressions which is useful for restricting parameters in non-ascii
character sets.

Refs #8332

markstory committed:
Merge pull request #8329 from ifunk/feature-assoc-classname

Add getter for protected $_className value for Association.

markstory committed:
Accept any DateTimeInterface as valid.

This fixes validators not accepting immutable datetime objects.

Refs #8333

lorenzo committed:
Merge pull request #8344 from cakephp/issue-8333

Accept any DateTimeInterface as valid.

lorenzo committed:
Merge pull request #8339 from cakephp/json-pretty-print

Make JsonConfig pretty print config files.

markstory committed:
Add multibytePattern option to Route.

Having an option to enable multibyte pattern support lets people opt-in
to more powerful but slower pattern matching. This also reduces the
chance that this new feature/option could cause issues as it is an
opt-in feature.

Refs #8332

markstory committed:
Merge pull request #8318 from cakephp/response-get-file

Add Response::getFile()

markstory committed:
Fix failing tests.

stmeyer commented:
The tests failed, due to a mistake in TravisCI (https://travis-ci.org/cakephp/cakephp/jobs/111167184#L185):
```
  [Composer\Downloader\TransportException]   
  Could not authenticate against github.com  
```
Is there a way to restart the checks without a new commit?

ADmad commented:
@stmeyer Restarted

Commit:
phpcs: change tab to space

lorenzo committed:
Merge pull request #8343 from cakephp/issue-8332

Make route patterns allow multibyte sub-patterns.

stmeyer commented:
@ADmad thanks :clap: 
@lorenzo just as information for future planning: when are you planning to create the new testsuit?

lorenzo commented:
I'm waiting for the weekend to work on this

lorenzo committed:
Merge pull request #8328 from cakephp/issue-8313

Quote table name in SQLServer schema.

stmeyer commented:
@lorenzo nice to hear :smile:

btaens committed:
Table mixin added to Association

markstory committed:
Merge pull request #8355 from btaens/master

Table mixin added to Association

markstory committed:
Attempt to clarify behavior of dependent option.

Refs #8348

ADmad committed:
Merge pull request #8358 from cakephp/dependent-docs

Attempt to clarify behavior of dependent option.

markstory committed:
Suppress notice errors in PO file parsing.

Suppress the notice errors when PO files include duplicate msgid
elements that have context and context-free variants. In these cases,
the context-free definition 'wins'.

Ideally we could add a '' context that is used by default. I tried this,
but it broke a significant number of tests which made me concerned about
backwards compatibility with userland message parsers.

Refs #8337

markstory committed:
Fix extracting POT files from vendor prefix plugins.

Correctly generate the output file name for vendor prefixed plugins.
As per the convention defined in MessagesFileLoader, prefixed plugins
only use the plugin name as the domain.po/pot file.

Refs #8316

markstory committed:
Handle negative numbers in Redis correctly.

Update number sniff to handle negative numbers. We need to do number
sniffing so we can maintain compatbility between write() and
increment()/decrement().

Refs #8364

ADmad committed:
Show warning if association property name is same as table field.

lorenzo committed:
Merge pull request #8363 from cakephp/issue-8316

Fix extracting POT files from vendor prefix plugins.

lorenzo committed:
Merge pull request #8361 from cakephp/issue-8337

Suppress notice errors in PO file parsing.

lorenzo committed:
Merge pull request #8366 from cakephp/issue-8364-3x

3.x - Handle negative numbers in Redis correctly.

antograssiot committed:
remove unsued import

antograssiot committed:
remove unused variables/lines

detected by scrutinizer

antograssiot committed:
fix argument number issue

ADmad committed:
Add tests for previous commit.

ADmad committed:
Reduce code duplication.

lorenzo committed:
Merge pull request #8370 from cakephp/namespace-cleanup

3.2 import cleanup in test suite

ADmad committed:
Remov newline from error message.

chinpei215 committed:
Using correct quotation marks.

lorenzo committed:
Merge pull request #8371 from chinpei215/correct-quotes

Using correct quotation marks.

lorenzo committed:
Improving code readability

lorenzo committed:
Improving code readability

lorenzo committed:
Removing extra space

markstory committed:
Merge pull request #8367 from cakephp/assoc-prop-name

Show warning if association property name is same as table field.

lorenzo committed:
Merge branch 'master' of https://github.com/stmeyer/cakephp into stmeyer-master

lorenzo commented:
I need to close this pull request and open it again with the code I added in and after rebasing the branch. It seems like it went crazy
